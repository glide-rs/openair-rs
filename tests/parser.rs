use std::io::Cursor;

use indoc::indoc;
use insta::assert_debug_snapshot;
use openair::*;

#[test]
fn test_switzerland_fixture() {
    let airspace = include_bytes!("../example_data/Switzerland.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

#[test]
fn test_germany_fixture() {
    let airspace = include_bytes!("../example_data/Germany.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

#[test]
fn test_germany_border_fixture() {
    let airspace = include_bytes!("../example_data/Germany_Border.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

/// Parse an airspace as generated by flyland.
#[test]
fn flyland_buochs() {
    let mut airspace = indoc! {"
        AC D
        AN BUOCHS Be CTR 119.625
        AL GND
        AH 12959 ft
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        DP 46:57:55 N 008:28:40 E
        DP 46:58:28 N 008:27:56 E
        DP 46:57:13 N 008:27:52 E
        * n-Points: 5
    "}
    .as_bytes();
    let mut spaces = parse(&mut airspace).unwrap();
    assert_eq!(spaces.len(), 1);
    let space: Airspace = spaces.pop().unwrap();
    assert_eq!(space.name, "BUOCHS Be CTR 119.625");
    assert_eq!(space.lower_bound, Altitude::Gnd);
    assert_eq!(space.upper_bound, Altitude::FeetAmsl(12959));
    if let Geometry::Polygon { segments } = space.geom {
        assert_eq!(segments.len(), 5);
    } else {
        panic!("Unexpected enum variant");
    }
}

/// The order of bounds should not matter.
#[test]
fn inverted_bounds() {
    let mut a1 = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH 12959 ft
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        *
    "}
    .as_bytes();
    let mut a2 = indoc! {"
        AC D
        AN SOMESPACE
        AH 12959 ft
        AL GND
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        *
    "}
    .as_bytes();
    let space1 = parse(&mut a1).unwrap().pop().unwrap();
    let space2 = parse(&mut a2).unwrap().pop().unwrap();
    assert_eq!(space1, space2);
}

/// Variables can be defined multiple times.
#[test]
fn multi_variable() {
    let mut a = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH FL100
        V X=52:00:00N 013:00:00E
        V D=+
        DA 2,0,30
        V X=52:00:00N 013:00:00E
        V D=-
        DA 4,60,30
        *
    "}
    .as_bytes();
    let airspace = parse(&mut a).unwrap().pop().unwrap();
    assert_eq!(
        airspace.geom,
        Geometry::Polygon {
            segments: vec![
                PolygonSegment::ArcSegment(ArcSegment {
                    centerpoint: Coord {
                        lat: 52.0,
                        lng: 13.0
                    },
                    radius: 2.0,
                    angle_start: 0.0,
                    angle_end: 30.0,
                    direction: Direction::Cw,
                }),
                PolygonSegment::ArcSegment(ArcSegment {
                    centerpoint: Coord {
                        lat: 52.0,
                        lng: 13.0
                    },
                    radius: 4.0,
                    angle_start: 60.0,
                    angle_end: 30.0,
                    direction: Direction::Ccw,
                }),
            ],
        }
    );
}

/// Test AY/AF/AG records.
#[test]
fn extension_records() {
    let mut a = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH 100 ft AGL
        AY AWY
        AF 132.350
        AG Dutch Mil
        AX 1234
        A* custom extension
        V X=52:00:00 N 013:00:00 E
        DC 5
    "}
    .as_bytes();
    let airspace = parse(&mut a).unwrap().pop().unwrap();
    assert_eq!(airspace.type_, Some("AWY".to_string()));
    assert_eq!(airspace.frequency, Some("132.350".to_string()));
    assert_eq!(airspace.call_sign, Some("Dutch Mil".to_string()));
    assert_eq!(airspace.transponder_code, Some(1234));
}
