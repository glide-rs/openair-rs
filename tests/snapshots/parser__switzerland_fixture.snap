---
source: tests/parser.rs
expression: spaces.first().unwrap()
---
Airspace {
    name: "ALPNACH CTR 128.475",
    class: D,
    type_: None,
    lower_bound: Gnd,
    upper_bound: FeetAmsl(
        12959,
    ),
    geom: Polygon {
        segments: [
            Point(
                Coord {
                    lat: 46.974722222222226,
                    lng: 8.258055555555556,
                },
            ),
            Point(
                Coord {
                    lat: 46.972500000000004,
                    lng: 8.279722222222222,
                },
            ),
            Point(
                Coord {
                    lat: 46.93666666666666,
                    lng: 8.267777777777779,
                },
            ),
            Point(
                Coord {
                    lat: 46.93444444444444,
                    lng: 8.245277777777776,
                },
            ),
            Point(
                Coord {
                    lat: 46.92638888888889,
                    lng: 8.226666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 46.926944444444445,
                    lng: 8.225833333333334,
                },
            ),
            Point(
                Coord {
                    lat: 46.92305555555555,
                    lng: 8.223611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 46.88972222222222,
                    lng: 8.241666666666665,
                },
            ),
            Point(
                Coord {
                    lat: 46.88333333333333,
                    lng: 8.2725,
                },
            ),
            Point(
                Coord {
                    lat: 46.88472222222222,
                    lng: 8.273055555555556,
                },
            ),
            Point(
                Coord {
                    lat: 46.88611111111111,
                    lng: 8.272777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 46.88805555555555,
                    lng: 8.270833333333334,
                },
            ),
            Point(
                Coord {
                    lat: 46.88916666666667,
                    lng: 8.270555555555557,
                },
            ),
            Point(
                Coord {
                    lat: 46.89111111111111,
                    lng: 8.2725,
                },
            ),
            Point(
                Coord {
                    lat: 46.8925,
                    lng: 8.271666666666668,
                },
            ),
            Point(
                Coord {
                    lat: 46.89416666666666,
                    lng: 8.273055555555556,
                },
            ),
            Point(
                Coord {
                    lat: 46.895833333333336,
                    lng: 8.273333333333333,
                },
            ),
            Point(
                Coord {
                    lat: 46.89972222222222,
                    lng: 8.275277777777779,
                },
            ),
            Point(
                Coord {
                    lat: 46.900555555555556,
                    lng: 8.275,
                },
            ),
            Point(
                Coord {
                    lat: 46.90222222222222,
                    lng: 8.276111111111112,
                },
            ),
            Point(
                Coord {
                    lat: 46.905,
                    lng: 8.285833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 46.9075,
                    lng: 8.290833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 46.91,
                    lng: 8.293611111111112,
                },
            ),
            Point(
                Coord {
                    lat: 46.91361111111111,
                    lng: 8.295833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 46.916666666666664,
                    lng: 8.296388888888888,
                },
            ),
            Point(
                Coord {
                    lat: 46.91888888888889,
                    lng: 8.29888888888889,
                },
            ),
            Point(
                Coord {
                    lat: 46.92055555555555,
                    lng: 8.3025,
                },
            ),
            Point(
                Coord {
                    lat: 46.92027777777778,
                    lng: 8.307222222222222,
                },
            ),
            Point(
                Coord {
                    lat: 46.920833333333334,
                    lng: 8.311944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 46.918055555555554,
                    lng: 8.317222222222222,
                },
            ),
            Point(
                Coord {
                    lat: 46.91777777777777,
                    lng: 8.321666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 46.9175,
                    lng: 8.323333333333332,
                },
            ),
            Point(
                Coord {
                    lat: 46.92944444444444,
                    lng: 8.340833333333334,
                },
            ),
            Point(
                Coord {
                    lat: 47.01027777777778,
                    lng: 8.309166666666668,
                },
            ),
            Point(
                Coord {
                    lat: 46.974722222222226,
                    lng: 8.258055555555556,
                },
            ),
        ],
    },
    frequency: None,
    call_sign: None,
    transponder_code: None,
}
